socket.go

```diff
@@ -561,123 +613,255 @@ func (socket *mongoSocket) Query(ops ...interface{}) (err error) {
 	return err
 }
 
-func fill(r net.Conn, b []byte) error {
-	l := len(b)
-	n, err := r.Read(b)
-	for n != l && err == nil {
-		var ni int
-		ni, err = r.Read(b[n:])
-		n += ni
+// sendMessage send data to the database using the OP_MSG wire protocol
+// introduced in MongoDB 3.6 (require maxWireVersion >= 6)
+func (socket *mongoSocket) sendMessage(op *msgOp) (writeCmdResult, error) {
+	var wr writeCmdResult
+	var err error
+
+	buf := bytesBufferPool.Get().([]byte)
+
+	buf = addHeader(buf, dbMessage)
+	buf = addInt32(buf, int32(op.flags))
+
+	for _, section := range op.sections {
+		buf, err = addSection(buf, section)
+		if err != nil {
+			return wr, err
+		}
 	}
-	return err
+
+	if len(buf) > socket.ServerInfo().MaxMessageSizeBytes {
+		return wr, fmt.Errorf("message length to long, should be < %v, but was %v", socket.ServerInfo().MaxMessageSizeBytes, len(buf))
```

Comment:
It seems this is possible to trigger by adding too many documents to the query batch?

We should probably either limit the user when building the batch (i.e. return an error to the caller of `Bulk.Insert()`) to avoid this fairly confusing error (we know what it is, but most won't). 

I'm not massively familiar with the bulk ops though, any ideas?

Reply:
`Insert()` currently doesn't return an error, so this would be a drastic change in the way Bulk operations works. Instead, I think the driver should handle this internally: 
for example, if someone tries to insert 100000 docs of 480+ bytes, the driver should split the operation in several smaller ops